### PR TITLE
TimeType subtraction using promote

### PR DESCRIPTION
**Pull request by baumgold**

### CODE REVIEW

**Review by quinnj — APPROVED** [review]

**Comment by Seelengrab** [comment]
I don't think this is necessarily a good change - conversion to a `DateTime` IMO should be explicit, and not implicitly default to midnight. Due to this PR, there also is now a worse error for `-(::Date, ::Time)`, which complains that "promotion failed". It would be better if it informed the user that they need to decide at what time their `Date` exists (I know that `Date + Time` already makes the same assumption, which I think is just as bad).

**Comment by baumgold** [comment]
> I don't think this is necessarily a good change - conversion to a `DateTime` IMO should be explicit, and not implicitly default to midnight. Due to this PR, there also is now a worse error for `-(::Date, ::Time)`, which complains that "promotion failed". It would be better if it informed the user that they need to decide at what time their `Date` exists (I know that `Date + Time` already makes the same assumption, which I think is just as bad).

That's fair.  Maybe this should be limited to `AbstractDateTime` rather than `TimeType`.  @Seelengrab - what do you think?

**Comment by Seelengrab** [comment]
I'm not sure. It depends on what you intended to do with the change originally - you gave no reasoning for it after all :shrug:

**Comment by baumgold** [comment]
Here’s my reasoning: I use an internal package that subtypes AbstractDateTime with some extra functionality not currently available in DateTime. I would like to be able to subtract DateTime and my custom internal type that subtypes AbstractDateTime. We might want this to be handled more generally for all TimeTypes, or may want to limit it to AbstractDateTime. I don’t have a strong opinion here.

**Comment by Seelengrab** [comment]
You're free to implement `-(::AbstractDateTime, ::MyDateTime)` on your type, as long as you use the API of `AbstractDateTime` to interact with it - that doesn't necessarily imply that `-(::AbstractDateTime, ::AbstractDateTime)` is good to have in Base (I don't think it is, for the reason stated above), but such a discussion should generally happen within a PR defining semantics before it's merged.

@quinnj can you clarify why you merged this? Do we have an official stance on which time a `Date` "sits" at? It's hard to follow this without a review comment..

**Comment by quinnj** [comment]
Why is it a problem to promote to midnight? Do you have any concrete examples where that would be "bad" by doing it implicitly and why it would be better to do it explicitly? I merged because I feel like I remember that we already have places where we do this kind of "midnight assumption" and I've never heard of it biting anyone or there being anything wrong in practice, so I'm just curious why the pushback? I'm certainly up for discussing if this is causing real problems for people somewhere.

**Comment by Seelengrab** [comment]
The issue is that in timezone aware code, there is no one "midnight", especially not in DST aware code, where there may be more than one midnight (or [no midnight at all](https://gist.github.com/timvisee/fcda9bbdff88d45cc9061606b4b923ca?permalink_comment_id=4275938#gistcomment-4275938) - the [whole list](https://gist.github.com/timvisee/fcda9bbdff88d45cc9061606b4b923ca) is a good source, and links to more sources).

I recognize that Base itself is oblivious to timezones (which I think is a bit of a mistake), but I think it would be a bigger mistake to encourage coding patterns that encourage that.

**Comment by Seelengrab** [comment]
All of that is fine and A-Ok - I'm not saying we should get rid of `Date`/`DateTime` entirely. It has its purpose! What I AM saying is that all this doesn't mean that we should encourage handling timezones/time conversions incorrectly/with more implicit assumptions. To that end, I think introducing codepaths that make assumptions such as "midnight will be fine" is a bad idea :tm: . That way lies future pain and complaints of footguns, which I'd rather like to avoid.

> And that's also why the package ecosystem is so great! You can have all these super-specific, nuanced Date packages that take all this stuff into account. Does the stdlib need improvement? Sure. And we should open specific issues for concrete things to improve. 

Great, let's discuss this in an issue then and revert this change for now, seeing as this wasn't discussed at all before being merged :) The original motivation of the PR can be done better by using dispatch after all, so that shouldn't block the development of a package.

**Comment by quinnj** [comment]
Sounds good; thanks.

**Comment by Seelengrab** [comment]
Done #49821 - should this PR be reverted directly, or should we just merge #49805, which seems much more reasonable/tame to me?

**Comment by baumgold** [comment]
I vote to merge #49805 but I’m okay reverting too if needed.
